README
NAME
    Mail::Chimp::API - Perl wrapper around the Mailchimp v1.1 API

SYNOPSIS
      use strict;
      use Mail::Chimp::API;

      my $chimp = Mail::Chimp::API->new($username, $password);

      my $campaigns   = $chimp->campaigns;
      my $lists       = $chimp->lists;
      my $subscribers = $chimp->listMembers($lists->[0]->{id});

      print 'success' if $chimp->listSubscribe($lists->[0]->{id}, 'someone@somewhere.com', {}, 1);

DESCRIPTION
    Mail::Chimp::API is a simple Perl wrapper around the MailChimp v1.1 API.
    The object exposes the MailChimp XML-RPC methods and confesses fault
    codes/messages when errors occur. Further, it keeps track of your API
    key for you so you do not need to enter it each time.

    Method parameters are passed straight through, in order, to MailChimp.
    Thus, you do need to understand the MailChimp v1.1 API as documented at
    <http://www.mailchimp.com/api/1.1/> so that you will know the
    appropriate parameters to pass to each method.

NOTES
    If you find yourself getting '-32602' errors, you are probably missing a
    required (even if empty) hash. The API documentation should be your
    first destination to verify the validity of your parameters. For
    example:

      # MailChimp Error -32602: server error. invalid method parameters
      print 'fail'    if $chimp->listSubscribe($lists->[0]->{id}, 'someone@somewhere.com');
  
      # The {} is required by MailChimp to indicate that you do not want any MergeVars
      print 'success' if $chimp->listSubscribe($lists->[0]->{id}, 'someone@somewhere.com', {}, 1);

    Also, the only security method implemented is login. This is likely to
    change at some point in the future.

DEPENDENCIES
      Mouse
      XMLRPC::Lite

SEE ALSO
      XMLRPC::Lite
      <http://www.mailchimp.com/api/1.1/>

COPYRIGHT
    Copyright (C) 2009 Dave Pirotte. All rights reserved.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    Dave Pirotte (dpirotte@gmail.com)



Changes
Revision history for Perl module Mail::Chimp

0.50
  - Add wrappers for all list related methods (Drew)
  - Switch to Moose (Drew)
  - Various bugfixes and tweaks (Gautam Dey)
  - Add support for v1.2 of the API (Ask)
  - Require api_version to be specified to new() to make
    upgrades easier / more reliable (Ask)
  - Add support for providing an api key instead of username
    and password (Ask Bjørn Hansen).
  - Begin creating wrapper methods for common API commands
    
0.12 April 11, 2009
  - Add more flexibility for API version (Drew Taylor)

0.11 March 15, 2009
  - First public release

0.01 Wed Mar 11 06:45:44 2009
    - original version; created by ExtUtils::ModuleMaker 0.51



lib/Mail/Chimp/List.pm
package Mail::Chimp::List;
use strict;
use warnings;
use Moose;
use MooseX::Types::DateTimeX qw(DateTime);
our $VERSION = '0.2.1';


has '_api'                         => (is => 'ro', isa => 'Mail::Chimp::API');
has 'id'                           => (is => 'ro', isa => 'Int');
has 'web_id'                       => (is => 'ro', isa => 'Int');
has 'name'                         => (is => 'ro', isa => 'Str');
has 'date_created'                 => (is => 'ro', isa => DateTime, coerce => 1);
has 'member_count'                 => (is => 'ro', isa => 'Int');
has 'unsubscribe_count'            => (is => 'ro', isa => 'Int');
has 'email_type_option'            => (is => 'ro', isa => 'Bool');
has 'default_from_name'            => (is => 'ro', isa => 'Str');
has 'default_from_email'           => (is => 'ro', isa => 'Str');
has 'default_subject'              => (is => 'ro', isa => 'Str');
has 'default_language'             => (is => 'ro', isa => 'Str');
has 'list_rating'                  => (is => 'ro', isa => 'Float');
has 'member_count_since_send'      => (is => 'ro', isa => 'Int');
has 'unsubscribe_since_send_count' => (is => 'ro', isa => 'Int');
has 'cleaned_count_since_send'     => (is => 'ro', isa => 'Int');


sub _call {
    my ( $self, $method, @args ) = @_;
    return $self->_api->_call( $method, $self->_api->apikey, $self->id, @args );
}

sub abuse_reports {
    my ( $self, $start, $limit, $since ) = @_;
    my $reports = $self->_call( 'listAbuseReports', $start, $limit, $since );
    return [ map { Mail::Chimp::AbuseReport->new( _api => $self->api, %$_) } @$reports ];
}

sub batch_subscribe {
    my ( $slf, $batch, $double_optin, $update_existing, $replace_interests ) = @_;
    return $self->_call( 'listBatchSubscribe', $batch, $double_optin, $update_existing, $replace_interests );
}

sub batch_unsubscribe {
    my ( $slf, $addresses, $delete_member, $send_goodbye, $send_notify ) = @_;
    return $self->_call( 'listBatchUnsubscribe', $addresses, $delete_member, $send_goodbye, $send_notify );
}

sub growth_history {
    my ( $self ) = @_;
    return $self->_call( 'listGrowthHistory' );
}

sub add_interest_group {
    my ( $self, $name ) = @_;
    return $self->_call( 'listInterestGroupAdd', $name );
}

sub delete_interest_group {
    my ( $self, $name ) = @_;
    return $self->_call( 'listInterestGroupDel', $name );
}

sub update_interest_group {
    my ( $self, $old_name, $new_name ) = @_;
    return $self->_call( 'listInterestGroupDel', $old_name, $new_name );
}

sub interest_groups {
    my ( $self ) = @_;
    return $self->_call( 'listInterestGroups' );
}

sub member_info {
    my ( $self, $email ) = @_;
    return $self->_call( 'listMemberInfo', $email );
}

sub members {
    my ( $self, $status, $since, $start, $limit ) = @_;
    return $self->_call( 'listMembers', $status, $since, $start, $limit );
}

sub add_merge_var {
    my ( $self, $name, $description, $options ) = @_;
    return $self->_call( 'listMergeVarAdd', uc $name, $description, $options );
}

sub delete_merge_var {
    my ( $self, $name ) = @_;
    return $self->_call( 'listMergeVarDel', uc $name );
}

sub update_merge_var {
    my ( $self, $name ) = @_;
    return $self->_call( 'listMergeVarUpdate', uc $name );
}

sub merge_vars {
    my ( $self ) = @_;
    return $self->_call( 'listMergeVars' );
}

sub subscribe_address {
    my ( $self, $email, $merge_vars ) = @_;
    $merge_vars ||= {};
    return $self->_call( 'listSubscribe', $email, $merge_vars );
}

sub unsubscribe_address {
    my ( $self, $email, $delete ) = @_;
    return $self->_call( 'listUnsubscribe', $email, $delete );
}

sub update_member {
    my ( $self, $email, $merge_vars, $email_type, $replace_interests ) = @_;
    $merge_vars ||= {};
    return $self->_call( 'listUpdateMember', $email, $merge_vars, $email_type, $replace_interests );
}

sub webhooks {
    my ( $self ) = @_;
    return $self->_call('listWebhooks' );
}

sub add_webhook {
    my ( $self, $url, $actions, $sources ) = @_;
    return $self->_call( 'listWebhookAdd', $url, $actions, $sources );
}

sub delete_webhook {
    my ( $self, $url, $actions, $sources ) = @_;
    return $self->_call( 'listWebhookDel', $url );
}

1;


lib/Mail/Chimp/API.pm
package Mail::Chimp::API;
use strict;
use warnings;
use Moose;
use XMLRPC::Lite;
use Data::Dumper;

our $VERSION = '0.2.1';


has 'username'      => (is => 'ro', isa => 'Str');
has 'password'      => (is => 'ro', isa => 'Str');
has 'api'           => (is => 'rw', isa => 'XMLRPC::Lite');
has 'api_version'   => (is => 'ro', isa => 'Num', default => 1.2);
has 'api_url'       => (is => 'ro', isa => 'Str');
has 'apikey'        => (is => 'rw', isa => 'Str');
has 'use_secure'    => (is => 'ro', isa => 'Bool', default => 1);
has 'datacenter'    => (is => 'ro', isa => 'Str', default => 'us1');
has 'output_format' => (is => 'rw', isa => 'Str', default => 'json');


sub BUILD {
    my ( $self ) = @_;
    
    die 'apikey or username and password is required'
        unless ($self->apikey or ($self->username and $self->password));
    my $url = 'http';
    $url .= 's' if $self->use_secure();
    $url .= "://" . $self->datacenter() . ".api.mailchimp.com/" . $self->api_version() . '/';
    warn "# $url\n";
    $self->api_url( $url );
    $self->api( XMLRPC::Lite->proxy( $url ) );
    unless ($self->apikey) {
        $self->apikey( $self->_call( 'login', $self->username(), $self->password() ) );
    }
}

sub _call {
    my $self = shift;

    my $call = $self->api->call( @_ );
    return $call->result
        || confess( sprintf( "MailChimp Error %d: %s", $call->fault->{faultCode}, $call->fault->{faultString} ) );
}

sub all_lists {
    my ( $self ) = @_;

    my $lists = $self->_call( 'lists', $self->apikey );
    require Mail::Chimp::List;
    return [ map { Mail::Chimp::List->new( _api => $self, %$_ ) } @$lists ];
}

sub get_list {
    my ( $self, $id ) = @_;
    my $lists = $self->get_lists();
    return grep { $_->id() == $id or $_->name() eq $id } @$lists;
}


{
    no strict 'refs';
    sub _make_api_method { 
        my ($class, $method) = @_;
        *{"${class}::$method"} = sub { my $self = shift; $self->_call( $method, $self->apikey, @_ ) };
    }
}

my @api_methods = qw(
    campaignContent
    campaignCreate
    campaignDelete
    campaignEcommAddOrder
    campaignFolders
    campaignPause
    campaignReplicate
    campaignResume
    campaignSchedule
    campaignSegmentTest
    campaignSendNow
    campaignSendTest
    campaignTemplates
    campaignUnschedule
    campaignUpdate
    campaigns

    campaignAbuseReports
    campaignAdvice
    campaignAnalytics
    campaignBounceMessages
    campaignClickStats
    campaignEcommOrders
    campaignEepUrlStats
    campaignEmailDomainPerformance
    campaignGeoOpens
    campaignGeoOpensForCountry
    campaignHardBounces
    campaignSoftBounces
    campaignStats
    campaignUnsubscribes

    campaignClickDetailAIM
    campaignEmailStatsAIM
    campaignEmailStatsAIMAll
    campaignNotOpenedAIM
    campaignOpenedAIM

    chimpChatter
    createFolder
    exommAddOrder
    generateText
    getAccountDetails
    getAffiliateInfo
    inlineCss
    listsForEmail
    ping

    listAbuseReports
    listBatchSubscribe
    listBatchUnsubscribe
    listGrowthHistory
    listInterestGroupAdd
    listInterestGroupDel
    listInterestGroupUpdate
    listInterestGroups
    listMemberInfo
    listMembers
    listMergeVarAdd
    listMergeVarDel
    listMergeVarUpdate
    listMergeVars
    listSubscribe
    listUnsubscribe
    listUpdateMember
    listWebhookAdd
    listWebhookDel
    listWebhooks
    lists
    
    apikeyAdd
    apikeyExpire
    apikeys
);

__PACKAGE__->_make_api_method( $_ ) for @api_methods;

=head1 NAME

Mail::Chimp::API - Perl wrapper around the Mailchimp API

=head1 SYNOPSIS

  use strict;
  use Mail::Chimp::API;

  my $chimp = Mail::Chimp::API->new(apikey      => $apikey,
                                    api_version => 1.2);

  # or if you have no apikey setup:
  # my $chimp = Mail::Chimp::API->new(username    => $username,
  #                                   password    => $password,
  #                                   api_version => 1.1,
  #                                  );

  my $apikey      = $chimp->apikey;   # generated on first login if none was setup

  my $campaigns   = $chimp->campaigns;
  my $lists       = $chimp->lists;
  my $subscribers = $chimp->listMembers($lists->[0]->{id});

  print 'success' if $chimp->listSubscribe($lists->[0]->{id}, 'someone@somewhere.com', {}, 1);

=head1 DESCRIPTION

Mail::Chimp::API is a simple Perl wrapper around the MailChimp API.
The object exposes the MailChimp XML-RPC methods and confesses
fault codes/messages when errors occur.  Further, it keeps track of
your API key for you so you do not need to enter it each time.

Method parameters are passed straight through, in order, to MailChimp.
Thus, you do need to understand the MailChimp API as documented
at <http://api.mailchimp.com/> so that you will know the
appropriate parameters to pass to each method.

This API has been tested with version 1.0, 1.1 and 1.2 of the API.

=head1 NOTES

If you find yourself getting '-32602' errors, you are probably missing
a required (even if empty) hash.  The API documentation should be your
first destination to verify the validity of your parameters.  For
example:

  # MailChimp Error -32602: server error. invalid method parameters
  print 'fail'    if $chimp->listSubscribe($lists->[0]->{id}, 'someone@somewhere.com');
  
  # The {} is required by MailChimp to indicate that you do not want any MergeVars
  print 'success' if $chimp->listSubscribe($lists->[0]->{id}, 'someone@somewhere.com', {}, 1);


=head1 DEPENDENCIES

  Moose
  XMLRPC::Lite

=head1 SEE ALSO

  XMLRPC::Lite
  <http://www.mailchimp.com/api/1.2/>

=head1 COPYRIGHT

Copyright (C) 2009 Dave Pirotte.  All rights reserved.

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=head1 AUTHOR

Dave Pirotte (dpirotte@gmail.com)

Drew Taylor (drew@drewtaylor.com)

Ask Bjørn Hansen (ask@develooper.com)

=cut

1;
